datastore/query.go

```diff
@@ -81,6 +92,145 @@ type order struct {
 	Direction sortDirection
 }
 
+// EntityFilter represents a datastore filter
+type EntityFilter interface {
+	toEntityFilter() (EntityFilter, error)
+	toProto() (*pb.Filter, error)
+}
+
+// PropertyFilter represent field based filter
+//
+// The operator parameter takes the following strings: ">", "<", ">=", "<=",
+// "=", "!=", "in", and "not-in".
+// Fields are compared against the provided value using the operator.
+// Field names which contain spaces, quote marks, or operator characters
+// should be passed as quoted Go string literals as returned by strconv.Quote
+// or the fmt package's %q verb.
+type PropertyFilter struct {
+	FieldName string
+	Operator  string
+	Value     interface{}
+}
+
+func (pf PropertyFilter) toProto() (*pb.Filter, error) {
+
+	if pf.FieldName == "" {
+		return nil, errors.New("datastore: empty query filter field name")
+	}
+	v, err := interfaceToProto(reflect.ValueOf(pf.Value).Interface(), false)
+	if err != nil {
+		return nil, fmt.Errorf("datastore: bad query filter value type: %v", err)
+	}
+	op, ok := operatorToProto[pf.Operator]
+	if !ok {
+		return nil, errors.New("datastore: unknown query filter operator")
+	}
+	xf := &pb.PropertyFilter{
+		Op:       op,
+		Property: &pb.PropertyReference{Name: pf.FieldName},
+		Value:    v,
+	}
+	return &pb.Filter{
+		FilterType: &pb.Filter_PropertyFilter{PropertyFilter: xf},
+	}, nil
+}
+
+func (pf PropertyFilter) toEntityFilter() (EntityFilter, error) {
+	op := strings.TrimSpace(pf.Operator)
+	_, isOp := operatorToProto[op]
+	if !isOp {
+		return nil, fmt.Errorf("datastore: invalid operator %q in filter", pf.Operator)
+	}
+	pf.Operator = op
+
+	unquotedFieldName, err := unquote(pf.FieldName)
+	if err != nil {
+		return nil, fmt.Errorf("datastore: invalid syntax for quoted field name %q", pf.FieldName)
+	}
+	pf.FieldName = unquotedFieldName
+	return pf, nil
+}
+
+// CompositeFilter represents datastore composite filters
+type CompositeFilter interface {
+	EntityFilter
+	isCompositeFilter()
+}
+
+// OR represents OR'ed filters
+type OR struct {
+	Filters []EntityFilter
+}
+
+func (OR) isCompositeFilter() {}
+
+func (or OR) toProto() (*pb.Filter, error) {
+
+	pbFilters := make([]*pb.Filter, len(or.Filters))
+
+	for i, filter := range or.Filters {
+		pbFilter, err := filter.toProto()
+		if err != nil {
+			return nil, err
+		}
+		pbFilters[i] = pbFilter
+	}
+	return &pb.Filter{FilterType: &pb.Filter_CompositeFilter{CompositeFilter: &pb.CompositeFilter{
+		Op:      pb.CompositeFilter_OR,
```

Comment:
Is this common naming convention in golang?

Reply:
Not really -- but there are all sorts of unusual naming conventions in protobufs.